### PR TITLE
[IMP] Mail: Move suggestion list code inside a model

### PR DESCRIPTION
**Pull request by phenix-factory**
This will reduce composer model complexity and allow to reuse suggestion code
inside other component.

task-2447140

### CODE REVIEW

File: addons/mail/static/src/models/suggestion_list/suggestion_list.js [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+import { registerNewModel } from '@mail/model/model_core';
+import { attr, many2many, many2one } from '@mail/model/model_field';
+import { link, unlink, unlinkAll } from '@mail/model/model_field_command';
+
+function factory(dependencies) {
+
+    class SuggestionList extends dependencies['mail.model'] {
+
+        closeSuggestions() {
+            this.update({ hasSuggestions: false });
```
Comment (seb-odoo) [inline addons/mail/static/src/models/suggestion_list/suggestion_list.js:12]: `hasSuggestions` should stay a compute on the lists having elements.

File: addons/mail/static/src/models/composer/composer.js [hunk 2]
```diff
@@ -51,6 +51,7 @@ function factory(dependencies) {
          */
         closeSuggestions() {
             this.update({ suggestionDelimiterPosition: clear() });
+            this.suggestionList.closeSuggestions();
```
Comment (seb-odoo) [inline addons/mail/static/src/models/composer/composer.js:54]: Removing the delimiter should clear the suggestion search term which should make the suggestion list empty.
It shouldn't be needed to call any manual "close" on it.

File: addons/mail/static/src/models/composer/composer.js [hunk 3]
```diff
@@ -799,27 +687,16 @@ function factory(dependencies) {
             mainSuggestedRecords.length = Math.min(mainSuggestedRecords.length, limit);
             extraSuggestedRecords.length = Math.min(extraSuggestedRecords.length, limit - mainSuggestedRecords.length);
             this.update({
-                extraSuggestedRecords: replace(extraSuggestedRecords),
                 hasToScrollToActiveSuggestion: true,
+            });
+            this.suggestionList.update({
```
Comment (seb-odoo) [inline addons/mail/static/src/models/composer/composer.js:692]: The whole code of `_updateSuggestionList` should be done inside the suggestion list itself.

The main interaction between composer and suggestion list should be that the composer is updating the search term.

**Review by seb-odoo — CHANGES_REQUESTED** [review]
It's better like this, but there is still an important part of suggestion list that should be moved into it, it is the search itself.

Not the delimiter part, but once the delimiter process has determined a model and search term it should be the list job to take it from there.

**Comment by phenix-factory** [comment]
The more I think of it the more I think this PR make no sense. It add complexity and solve 0 problem.
It will never reach the state where, to support suggestions, we just wrap an input in a suggestion_list component and it will work.

**Comment by seb-odoo** [comment]
> The more I think of it the more I think this PR make no sense. It add complexity and solve 0 problem.
> It will never reach the state where, to support suggestions, we just wrap an input in a suggestion_list component and it will work.

It's probably correct to say this goal will never be reached, and it's fine. I don't think it is adding complexity, it should make the code easier to maintain by correctly splitting concerns. If you don't have this feeling yet, it means the PR is not good enough yet.

**Pull request closed without merging**